simvacation/simunvacation.8
.TH SIMUNVACATION "8" "22 February 2005" "U-M Directory Project" "MAINTENANCE COMMANDS"
.SH NAME
.B simunvacation
\- remove unneeded database files generated by the
.B simvacation(8)
program.
.SH SYNOPSIS
.B simunvacation
[
.B \-v
.I database-directory
]
[
.B \-s
.I searchbase
]
[
.B \-h
.I ldap-host
]
[
.B \-p
.I ldap-port
]
.SH DESCRIPTION
.B
simunvacation
removes any database files in
.I database-directory
which are no longer needed by the
.B simvacation(8)
program.  The database files are deemed unneeded
if (1) the user's onVacation attribute is set to FALSE or has no
value, or (2) if the user does not appear in the LDAP directory.
.LP
.B
simunvacation
should be run at least nightly via cron.
.LP
.B
simunvacation
logs its activity via syslog(3), with facility LOG_MAIL.
.SH OPTIONS
.TP
.B \-v
Specify the location of the database file directory.  Default is
/var/vacationdb.
.TP
.B \-h
Specify the host running the LDAP (Lightweight Directory Access Protocol)
daemon.
.TP
.B \-p
Specify the tcp port to use when connecting to the LDAP server.
.SH FILES
.B simunvacation
cleans up database files named
.B $USER.ddb
in
.B /var/vacationdb/[a-z]/.
.SH SEE ALSO
.BR simvacation(8),
.BR simvacreport(8),
.BR sendmail(8),
.BR ldap(3)
.LP
.B RFC 2251
\- Lightweight Directory Access Protocol (v3)
.LP


CHANGELOG
0.8.3.c
    - BOUNCE_JAIL directive added
    - debugging statements added
    - jail debugged
    - changed BOUNCE_SECONDS to mean 0 is forever

0.8.3.b
    - many bugfixes
    - queue scheduler bug
    - envelope API changes needed to support MID and sender lists
    - simc -d changes simta's debug level
    - simta_openlog() used instead of openlog()
    - MAIL_JAIL directive added
    - simc works for MID, queue, and sender
    - DISK_FACTOR and COMMAND_FACTOR added
    - GLOBAL_RELAY_DISABLE_SENDER_CHECKING -> DISABLE_SENDER_CHECKING

0.8.3.a
    - simta one process model
    - simta_gettimeofday() preffered interface

0.8.2.4
    - timeout bugfixes

0.8.2.4
    - debug logging added
    - nomenclature changes

0.8.2.3
    - bugfixes

0.8.2.2
    - bugfixes

0.8.2.1
    - DEBUG_LOGGING directive added to help debug the stuck message bugfix
    - log transition to DATA for outbounder debugging
    - E file format change for queue prioritization
    - additional logging for host deliver ordering
    - delivery activity logs in all cases
    - additional logging to help track down "no deliver" poison message
    - logging for SASL bug @ DEBUG_LOGGING level 1
    - receive inbound timer overhaul, add outbound timers


0.8.2.d
    - ignore trailing dot in email addresses
    - removed timer debugging logging
    - disable from checking a global option now
    - DELIVER_AFTER_ACCEPT gos away

0.8.2.c
    - global_relay does normal domain checking for "Mail from"
    - timer bugfixes
    - RBL_LOG_ONLY bugfix

0.8.2.b
    - RBL_LOG_ONLY now a valid action
    - DNS string progression better defined, content filter defs changed

0.8.2.a
    - global connection throttles
    - connect startup sequence audit
    - local throttles overridden by RBL ACCEPT
    - connect sequence change
    - client hostname is "Unresolved" before DNS check
    - RBL check more error tolerant
    - CF Bounce
    - JAIL bugfix
    - REFUSE_MODE bugfix
    - DELIVER_AFTER_ACCEPT
    - timers

0.8.1.ab
    - simsendmail -r is the same as -f
    - simsendmail generates FROM header based on envelope from address
    - corrected Receive syslogs in header.c
    - hardfail message too large and too many received headers

0.8.1.aa
    - dont overwrite CID for cf

0.8.1.z
    - bad return bugfix

0.8.1.y
    - Don't hang up after 554

0.8.1.x
    - MAX_CONNECT log bug fixed

0.8.1.w
    - make sure that RBL_BLOCK gives back IP Address
    - bugfix for TIMEOUT call
    - more standardization for smtp banners

0.8.1.v
    - SMTP_OFF mode defined
    - TCP Wrapers before RBLs
    - RBL bugfix
    - string mismatch bugfix

0.8.1.u
    - lbwrap_url bugfix
    - banner audit of 421/251

0.8.1.t
    - transmitted size bugfix

0.8.1.s
    - queue time bug fixed
    - SIMTA_PID, SIMTA_CID variables in cf now
    - report data read and wrote to a client relative to Dfile size
    - no double syslog message
    - log more of what the client says

0.8.1.r
    - bugfix

0.8.1.q
    - better rcpt accounting/logging
    - smtp functions returning SYSERROR get proper SMTP reply

0.8.1.p
    - better mail from/rcpt to metric counting
    - WRITE_BEFORE_BANNER config file bugfix

0.8.1.o
    - bugfixes

0.8.1.n
    - consuistent metrics in tempfail mode
    - SMTP mode changes are logged
    - Receive logging standard adhered to

0.8.1.m
    - Auth logging fixed
    - RBL logging fixed

0.8.1.l
    - makefile patch
    - RBL "secret messages" bugfix
    - SIMTA_UID available in the content filter
    - RBL accounting bugfig
    - switched from double to single linked list for connection_info tracking
    - syslog openlog() bug fixed

0.8.1.k
    - bugfix for connection search uninitilized variable fixed

0.8.1.j
    - simta unique PIDs
    - RBL logging cleaned up
    - connection throttling on a per IP basis introduced

0.8.1.i
    - pidfd locking hopefully better
    - Timezone info added to Received header
    - logging Members Only accept/deny explained
    - logging expansion depth metric
    - bugfix for sender matching memonly group

0.8.1.h
    - MAXHOSTNAMELEN bug fixed
    - RBL logging now includes failure/accept codes
    - better receive logging
    - DATA_SESSION_WAIT added to simta.c
    - bug: JAIL_HOST bugfixes

0.8.1.g
    - logging bugfixes

0.8.1.f
    - snet bugfixes

0.8.1.e
    - bugfix

0.8.1.d
    - content filter now logs who AUTHed

0.8.1.c
    - f_data better optimized for content filter
    - better error propagation

0.8.1.b
    - bugfix TARPIT rcpt to always accepts
    - bugfix for content filter illegal accept return in tarpit mode
    - bugfixes for libsnet
    - log change for base64 encode errors
    - receive always checks for write before banner

0.8.1.a
    - content filter message propogated to SMTP client
    - LIBWRAP_URL directive added
    - content filter check for write before banner happens on all connections
    - bugfix for RBL disconnect
    - content filter supports mail jail
    - content filter no longer opens Dfile on stdin
    - content filter envoked in every delivery mode
    - content filter return value now a bitfield
    - content filter supports DISCONNECT

0.8.0.6
    - tarpit_sleep() optimized

0.8.0.5
    - bugfix: return SMTP mode to the user default mode after auth

0.8.0.4
    - bugfix: f_data connection drop after message acceptance error

0.8.0.3
    - multiple Queue Added bugfix

0.8.0.2
    - bitbucket bugfix

0.8.0.1
    - simsendmail logs message size
    - bitbucket delivery option for deliver testing
    - better Daemon logging
    - local queue runner accounting bugfix

0.8.0
    - SMTP failed rcpt punishment bugfix
    - DISCONNECT is a possible failed rcpt punishment
    - faster disk read times
    - outbound SMTP DNS has better logging and is more consice
    - better process accounting, and logging
    - better listen connection management
    - consolidate envelope reading functions, and make mismatches non-fatal
    - uninitilized variable bugfix

simvacation/simunvacation.c
/*
 * unvacate.c - clean up database files for people who are no longer
 * "on vacation".
 *
 * Copyright (c) 2004 Regents of The University of Michigan.
 * All Rights Reserved.  See COPYRIGHT.
 */

#include <stdio.h>
#include <dirent.h>
#include <sys/time.h>
#include <sys/param.h>
#include <syslog.h>
#include <string.h>
#include <unistd.h>

#include <lber.h>
#include <ldap.h>

#ifndef VDBDIR
#define VDBDIR		"/var/vacationdb" /* dir for vacation databases */
#endif

#define	LDAP_HOST	"ldap.itd.umich.edu"
#define	BIND_DN		NULL
#define BIND_METHOD	NULL
#define SEARCHBASE	"ou=People,dc=umich,dc=edu"
#define	ATTR_ONVAC	"onVacation"
#define ATTRS		{ ATTR_ONVAC, NULL }

typedef struct uniq_list {
    char *un_name;
    struct uniq_list *un_next;
} uniq_list;

main( int argc, char **argv)
{
    extern int optind, opterr;
    extern char *optarg;

    char ch;

    LDAP *ld;
    char *ldap_host = LDAP_HOST;
    char *searchbase = SEARCHBASE;
    int ldap_port = LDAP_PORT;

    int rc;
    int err = 0;
    static struct timeval timeout;
    static char *attrs[] = ATTRS;
    char filter[64];LDAPMessage *result, *e;
    char **vac;

    char *vdbpath = VDBDIR;
    char vdbdir[MAXPATHLEN];
    char buf[MAXPATHLEN];
    int matches;
    DIR *dirp;
    struct dirent *dire;
    struct uniq_list *u, *uniqnames = NULL;
    struct uniq_list **unp = &uniqnames;
    char *perr;

    openlog( "unvacate", LOG_PID, LOG_VACATION );
    /*
     * Timeout values for LDAP searches
     */
    timeout.tv_sec = 30L;
    timeout.tv_usec = 0L;

    while (( ch = getopt( argc, argv, "s:h:p:v:" )) != EOF ) {
	switch( (char) ch ) {
	case 's':		/* seachbase */
	    searchbase = optarg;
	    break;

	case 'h':		/* ldap server */
	    ldap_host = optarg;
	    break;

	case 'p':		/* ldap port */
	    if ( isdigit( *optarg )) {
		ldap_port = atoi( optarg );
	    }
	    break;

	case 'v':		/* dir for vacation db files */
	    vdbpath = optarg;
	    break;

	case '?':
	default:
	    usage();
	}
    }

    /*
     * Connect to ldap server and bind
     */
    if (( ld = ldap_open( ldap_host, ldap_port )) == NULL ) {
	syslog( LOG_ALERT, "cannot conenct to host %s, port %d",
		ldap_host, ldap_port );
	exit( 1 );
    }
    if (( rc = ldap_simple_bind_s( ld, BIND_DN, BIND_METHOD ))
	  != LDAP_SUCCESS ) {
	ldap_get_option ( ld, LDAP_OPT_ERROR_STRING, &perr);
	syslog( LOG_ALERT, "ldap_simple_bind failed: %s",perr);
	free (perr);
	exit( 1 );
    }


    if ( access( vdbpath, F_OK )) {
	syslog( LOG_ALERT, "cannot open directory %s", vdbpath );
	exit( 1 );
    }

    /*
     * Build a linked list of candidate uniqnames.
     */
    for ( ch = 'a'; ch <= 'z'; ch++ ) {
	sprintf( vdbdir, "%s/%c", vdbpath, ch );
	if ((dirp = opendir( vdbdir )) == NULL) {
	    syslog( LOG_ALERT, "cannot open directory %s", vdbdir );
	    err++;
	    continue;
	}

	while (( dire = readdir( dirp )) != NULL ) {
	    int		len;

	    /* is this a dbm file? */

	    len = strlen( dire->d_name );
	    if ( len > 4 ) {
		if ( !strncmp( &dire->d_name[len - 4],
			      ".ddb", 4 )) {
		    strncpy( buf, dire->d_name, len - 4);
		    buf[len - 4] = '\0';
		    ( *unp ) = ( struct uniq_list *)
			       malloc( sizeof ( struct uniq_list ));
		    ( *unp )->un_name = strdup( buf );
		    unp = &(( *unp )->un_next );
		}
	    }
	}
    }
    ( *unp ) = NULL;

    /*
     * For each candidate uniqname, check to see if user is
     * still on vacation.  If not, remove the .dir and .pag files.
     */
     for ( u = uniqnames; u; u = u->un_next ) {
	sprintf( filter, "uid=%s", u->un_name );
	rc = ldap_search_st( ld, searchbase, LDAP_SCOPE_SUBTREE,
			     filter, attrs, 0, &timeout, &result );
	if ( rc != LDAP_SUCCESS ) {
#if 0
	    syslog( LOG_ALERT, "error in ldap_search: %s",
		    ldap_err2string( ld->ld_errno ));
#endif
	    ldap_get_option ( ld, LDAP_OPT_ERROR_STRING, &perr);
	    syslog( LOG_ALERT, "error in ldap_search: %s",perr);
	    free (perr);
	    err++;
	    continue;
	}
	matches = ldap_count_entries( ld, result );
	if ( matches > 1 ) {
	    syslog( LOG_ALERT, "ambiguous: %s", u->un_name );
	    err++;
	} else {
	    if ( matches == 0 ) {
		/* I don't consider this an error */
		syslog( LOG_INFO, "%s not in UMOD", u->un_name );
	    } else {
		e = ldap_first_entry( ld, result );
		vac = ldap_get_values( ld, e, ATTR_ONVAC );
	    }

	    /*
	     * If (1) user was not in LDAP Directory, or (2) no "onVacation"
	     * attribute was retrieved, or (3) the user is not on
	     * vacation, remove the dbm files.
	     */
	    if ( !matches || !vac || strcasecmp( *vac, "TRUE" )) {
		syslog( LOG_INFO, "removing %s's db files", u->un_name );
		sprintf( buf, "%s/%c/%s.ddb", vdbpath, u->un_name[0],
			 u->un_name );
		if ( unlink( buf ) < 0 ) {
		    syslog( LOG_ALERT, "cannot remove file %s (%m)", buf );
		    err++;
		}
	    } else {
		syslog( LOG_DEBUG, "user %s still on vacation", u->un_name );
	    }
	}
    }
    if ( err ) {
	exit( 1 );
    } else {
	exit( 0 );
    }
}

usage()
{
    fprintf( stderr, "usage: unvacate [-v vbdir] [-s searchbase]\n" );
    fprintf( stderr, "                [-h ldap_host] [-p ldap_port]\n" );
    exit( 1 );
}
